tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp

```diff
@@ -877,14 +878,47 @@ getClangDeclarationName(clang::ASTContext &Ctx, NameTranslatingInfo &Info) {
   }
 }
 
+static DeclName
+getSwiftDeclName(ASTContext &Ctx, NameTranslatingInfo &Info) {
+  assert(SwiftLangSupport::getNameKindForUID(Info.NameKind) == NameKind::Swift);
+  std::vector<Identifier> Args(Info.ArgNames.size(), Identifier());
+  std::transform(Info.ArgNames.begin(), Info.ArgNames.end(), Args.begin(),
+                 [&](StringRef T) { return Ctx.getIdentifier(T); });
+  return DeclName(Ctx, Ctx.getIdentifier(Info.BaseName),
+                  llvm::makeArrayRef(Args));
+}
+
 /// Returns true for failure to resolve.
 static bool passNameInfoForDecl(const ValueDecl *VD, NameTranslatingInfo &Info,
                     std::function<void(const NameTranslatingInfo &)> Receiver) {
   switch (SwiftLangSupport::getNameKindForUID(Info.NameKind)) {
   case NameKind::Swift: {
-
-    // FIXME: Implement the swift to objc name translation.
-    return true;
+    NameTranslatingInfo Result;
+    auto &Ctx = VD->getDeclContext()->getASTContext();
+    auto ResultPair = getObjCNameForSwiftDecl(VD, getSwiftDeclName(Ctx, Info));
+    Identifier Name = ResultPair.first;
+    if (!Name.empty()) {
+      Result.NameKind = SwiftLangSupport::getUIDForNameKind(NameKind::ObjC);
+      Result.BaseName = Name.str();
+    } else if (ObjCSelector Selector = ResultPair.second) {
+      Result.NameKind = SwiftLangSupport::getUIDForNameKind(NameKind::ObjC);
+      SmallString<64> Buffer;
+      StringRef Total = Selector.getString(Buffer);
+      SmallVector<StringRef, 4> Pieces;
+      Total.split(Pieces, ":");
+      if (Selector.getNumArgs()) {
+        assert(Pieces.back().empty());
+        Pieces.pop_back();
+        std::transform(Pieces.begin(), Pieces.end(), Pieces.begin(),
+          [](StringRef P) { return StringRef(P.data(), P.size() + 1); });
+      }
+      Result.ArgNames = llvm::makeArrayRef(Pieces);
```

Comment:
This looks suspicious, even though I know ArgNames isn't an ArrayRef itself. Can you go through a different API just to make it not look like you're storing the ArrayRef?